src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -203,8 +203,9 @@ trait TypeAssigner {
    */
   def selectionType(site: Type, name: Name, pos: Position)(implicit ctx: Context): Type = {
     val mbr = site.member(name)
-    if (reallyExists(mbr)) site.select(name, mbr)
-    else if (site.derivesFrom(defn.DynamicClass) && !Dynamic.isDynamicMethod(name)) {
+    lazy val canBeDynamicMethod = site.derivesFrom(defn.DynamicClass) && !Dynamic.isDynamicMethod(name)
+    if (reallyExists(mbr) && (mbr.accessibleFrom(site).exists || !canBeDynamicMethod)) site.select(name, mbr)
```

Comment:
No, a private def in a parent will not show up as a member of a subclass. But there could be another accessibility problem, that's true. Still, I think it's wrong to put accessibility concerns here. That's what we have function accessibleSelectionType for, after all. It's illogical that accessibility enters into selectionType logic for Dynamic resolution but not in other ways. To give a comparison: Here's what happens with implicit search:

    scala> class C { protected def foo: Int = ??? } 
    defined class C
    scala> val c = new C 
    c: C = C@68f4865
    scala> c.foo 
    <console>:7: error: method foo in class C cannot be accessed as a member of C(c) from module class $iw$. Access to protected method foo not permitted because
     enclosing object $iw is not a subclass of
     class C where target is defined
    c.foo
      ^
    scala> class D { def foo: Int = ??? } 
    defined class D
    scala> implicit def c2d(x: C): D = new D 
    c2d: (x: C)D
    scala> c.foo 
    <console>:9: error: method foo in class C cannot be accessed as a member of C(c) from module class $iw$. Access to protected method foo not permitted because
     enclosing object $iw is not a subclass of
     class C where target is defined
    c.foo
      ^
    scala> c2d(c).foo 
    scala.NotImplementedError: an implementation is missing

As you see, the implicit is _not_ tried if a normal member is found but that member is inaccessible. I believe Dynamic should work the same way.

Keep in mind in all this that the concept of private is not related to the concept of accessibility (in either Scala or Java). It looks like it, but it isn't. A private definition is simply not a member in a subclass. Accessibility does not enter the picture here.